python/plugins/processing/algs/qgis/RasterCalculator.py

```diff
@@ -122,18 +107,19 @@ def processAlgorithm(self, parameters, context, feedback):
 
         layersDict = {}
         if layers:
-            layersDict = {os.path.basename(lyr.source().split(".")[0]): lyr for lyr in layers}
+            layersDict = {lyr.source(): lyr for lyr in layers}
 
         crs = self.parameterAsCrs(parameters, self.CRS, context)
-        if not layers and not crs.isValid():
-            raise QgsProcessingException(self.tr("No reference layer selected nor CRS provided"))
-
-        if not crs.isValid() and layers:
-            crs = list(layersDict.values())[0].crs()
+        if crs is None or not crs.isValid():
+            if not layers:
+                raise QgsProcessingException(self.tr("No reference layer selected nor CRS provided"))
+            else:
+                crs = list(layersDict.values())[0].crs()
 
         bbox = self.parameterAsExtent(parameters, self.EXTENT, context)
-        if not layers and bbox.isNull():
-            raise QgsProcessingException(self.tr("No reference layer selected nor extent box provided"))
+        if bbox.isNull():
+            if not layers:
```

Comment:
To me this is less readable than the previous style, but maybe this is a matter of taste?

Reply:
because layers array can be populated or not (depending if set) and bbox can be set independently by a layer reference.

Reply:
ops... right